src/python_testing/TC_CGEN_2_2.py

```diff
@@ -0,0 +1,873 @@
+        th2_csr = await self.send_single_cmd(dev_ctrl=TH2, node_id=newNodeId, cmd=cmd)
+
+        # Flow generates a new TrustedRootCertificate - Isue the certificates
+        th2_certs_new = await TH2.IssueNOCChain(th2_csr, newNodeId)
+        th2_new_root_cert = th2_certs_new.rcacBytes
+
+        # Flow generates a new TrustedRootCertificate - Send command to add new trusted root certificate
+        cmd = self.cluster_opcreds.Commands.AddTrustedRootCertificate(th2_new_root_cert)
+        resp = await self.send_single_cmd(dev_ctrl=TH2, node_id=newNodeId, cmd=cmd)
+
+        self.step(17)
+        cmd = Clusters.GeneralCommissioning.Commands.ArmFailSafe(expiryLengthSeconds=0)
+        resp = await self.send_single_cmd(
+            dev_ctrl=TH2,
+            node_id=newNodeId,
+            cmd=cmd)
+        # Verify that the DUT responds with ArmFailSafeResponse with ErrorCode as 'OK'(0)
+        asserts.assert_equal(resp.errorCode, Clusters.GeneralCommissioning.Enums.CommissioningErrorEnum.kOk,
+                             "Failure status returned from arm failsafe")
+        # Verify that DebugText is empty or has a maximum length of 512 characters
+        debug_text = resp.debugText
+        asserts.assert_true(debug_text == '' or len(debug_text) <= 512,
+                            "debugText must be empty or have a maximum length of 512 characters.")
+        logger.info(f'Step #17 - ArmFailSafeResponse with ErrorCode as OK({resp.errorCode})')
+
+        self.step(18)
+        nocs = await self.read_single_attribute_check_success(
+            dev_ctrl=self.default_controller,
+            node_id=self.dut_node_id,
+            cluster=self.cluster_opcreds,
+            attribute=self.cluster_opcreds.Attributes.NOCs,
+            fabric_filtered=False)
+        nocs_original_size = len(nocs)
+        logger.info(f'Step #18 - TH1 Original size of the nocs list: {nocs_original_size}')
+
+        self.step(19)
+        fabrics = await self.read_single_attribute_check_success(
+            dev_ctrl=self.default_controller,
+            node_id=self.dut_node_id,
+            cluster=self.cluster_opcreds,
+            attribute=self.cluster_opcreds.Attributes.Fabrics,
+            fabric_filtered=False)
+        fabrics_original_size = len(fabrics)
+        logger.info(f'Step #19 - TH1 Original size of the fabrics list: {fabrics_original_size}')
+
+        self.step(20)
+        trusted_roots_list = await self.read_single_attribute_check_success(
+            dev_ctrl=self.default_controller,
+            node_id=self.dut_node_id,
+            cluster=self.cluster_opcreds,
+            attribute=self.cluster_opcreds.Attributes.TrustedRootCertificates)
+        trusted_roots_list_size = len(trusted_roots_list)
+        logger.info(f'Step #20 - TH1 Original size of the trusted_root list: {trusted_roots_list_size}')
+
+        self.step(21)
+        # Commissioning stage numbers - we should find a better way to match these to the C++ code
```

Comment:
It looks like the folks doing some of the new features plumbed through the functions required to do this using the commissioning parameters. Can you take a look at CGEN-2.5 (I James used it a few places). https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/TC_CGEN_2_5.py#L88. That will make this much neater, and I think it means you don't need to use the test commissioner.